File: render-javafx/src/main/java/dev/kkorolyov/pancake/render/javafx/JavaFXBox.java
package dev.kkorolyov.pancake.render.javafx;

import dev.kkorolyov.pancake.platform.math.Vector;
import dev.kkorolyov.pancake.platform.media.graphic.RenderTransform;
import dev.kkorolyov.pancake.platform.media.graphic.shape.Box;

import javafx.scene.canvas.GraphicsContext;

/**
 * {@link Box} implemented through JavaFX.
 */
public class JavaFXBox extends Box {
	private final EnhancedGraphicsContext g;
	private final Vector finalSize = new Vector();

	/**
	 * Constructs a new JavaFX-backed box instance.
	 * @param g backing graphics context
	 */
	public JavaFXBox(EnhancedGraphicsContext g) {
		this.g = g;
	}

	@Override
	public void render(RenderTransform transform) {
		finalSize
				.set(getSize())
				.scale(transform.getScale());

		GraphicsContext gg = g
				.shape(this)
				.rotate(transform)
				.get();

		gg.fillRect(
				transform.getPosition().getX(),
				transform.getPosition().getY(),
				finalSize.getX(),
				finalSize.getY()
		);
		gg.strokeRect(
				transform.getPosition().getX(),
				transform.getPosition().getY(),
				finalSize.getX(),
				finalSize.getY()
		);
	}
}

File: javafx-application/src/main/kotlin/dev/kkorolyov/pancake/javafx/render/JavaFxBox.kt
package dev.kkorolyov.pancake.javafx.render

import dev.kkorolyov.pancake.platform.math.Vector3
import dev.kkorolyov.pancake.platform.math.Vectors
import dev.kkorolyov.pancake.platform.media.graphic.RenderTransform
import dev.kkorolyov.pancake.platform.media.graphic.shape.Box

/**
 * [Box] implemented through JavaFX.
 */
class JavaFxBox(private val unitPixels: Double, private val g: EnhancedGraphicsContext) : Box() {
	private val finalSize: Vector3 = Vectors.create(0.0, 0.0, 0.0)

	override fun render(transform: RenderTransform) {
		finalSize.apply {
			set(size)
			scale(unitPixels)

			x *= transform.scale.x
			y *= transform.scale.y
		}

		g.run {
			shape(this@JavaFxBox)
			rotate(transform)

			get().run {
				fillRect(
					transform.position.x - (finalSize.x / 2),
					transform.position.y - (finalSize.y / 2),
					finalSize.x,
					finalSize.y
				)
				strokeRect(
					transform.position.x - (finalSize.x / 2),
					transform.position.y - (finalSize.y / 2),
					finalSize.x,
					finalSize.y
				)
			}
		}
	}
}
